core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -1790,17 +1790,41 @@ class KafkaApis(val requestChannel: RequestChannel,
       else {
         val supportedFeatures = brokerFeatures.supportedFeatures
         val finalizedFeaturesOpt = finalizedFeatureCache.get
-        finalizedFeaturesOpt match {
-          case Some(finalizedFeatures) => ApiVersion.apiVersionsResponse(
-            requestThrottleMs,
-            config.interBrokerProtocolVersion.recordVersion.value,
-            supportedFeatures,
-            finalizedFeatures.features,
-            finalizedFeatures.epoch)
-          case None => ApiVersion.apiVersionsResponse(
-            requestThrottleMs,
-            config.interBrokerProtocolVersion.recordVersion.value,
-            supportedFeatures)
+        val controllerApiVersions = if (isForwardingEnabled(request)) {
+          forwardingManager.controllerApiVersions()
+        } else
+          None
+
+        if (isForwardingEnabled(request) && controllerApiVersions.isEmpty) {
```

Comment:
This seems a little severe. Not all apis need to be coordinated with the controller. We have to handle the case when we receive a version that the current controller cannot handle anyway, so I think it's ok to make this a best-effort intersection and return just the broker APIs if the controller APIs are not yet known.

The tricky case for this PR is when the controller changed or we learned about new API version support after a client had already connected to the broker and sent `ApiVersions`. In this case, we have to detect version incompatibility dynamically when we try to forward the request. I might be missing something, but the current patch doesn't seem to handle this. Maybe the simplest option is to let the controller return UNSUPPORTED_VERSION in the envelope response if the header indicates an api or version that it does not support. Then the question is whether this error should be sent back to the client, but that would be a little surprising. 

Consider this sequence:

1. Client connects and sends ApiVersions request
2. Current controller supports AlterConfig v0-4, so that is what the broker indicates in the ApiVersions response
3. New controller is elected and only supports AlterConfig v0-3
4. Client sends AlterConfig v4

Now what happens? An unsupported version error here would be treated as fatal by the client. Instead, I think we agreed that instead of sending back the error, the broker would just disconnect. This would force a reconnect and a refresh of the API version support.

Reply:
I think it relates to the timing of the forwarding. When the controller switches and certain ApiVersions are updated, the forwarding response will contain unsupported version exception. At that time, we would do a disconnect to the client. We have two options here to trigger the disconnect:

1.  when we detect unsupported version exception from the forwarding response
2. when we detect the new ApiVersion set is different from the previous one

I personally think #1 is more accurate, but it has a downside where we couldn't tell whether it was truly an incompatible admin client trying to make connection. #2 may have wider impact than necessary, although it could disconnect clients with unaffected RPCs unnecessarily.